src/git/repo.rs

```diff
@@ -1049,6 +1086,17 @@ impl Index {
             file_mode: entry.mode,
         })
     }
+
+    pub fn update_all(&mut self, pathspecs: &[&str]) -> eyre::Result<()> {
+        self.inner.update_all(pathspecs, None)?;
+        Ok(())
+    }
+
+    pub fn write_tree(&mut self) -> eyre::Result<NonZeroOid> {
```

Comment:
We can't afford to write the entire index as a tree for performance reasons. To test performance, try running on https://github.com/mozilla/gecko-dev. You can also add a benchmark to `benches.rs` to confirm performance differences between implementations.

See the commentary and implementation for `Repo::cherry_pick_fast`. You have to take an existing tree and then augment it with only the changed paths, to ensure that we don't spend a lot of time doing unnecessary disk I/O:

https://github.com/arxanas/git-branchless/blob/331b7cf2a37d00a3d54ba5df00f3d702aa7b3948/src/git/repo.rs#L774-L789

Reply:
That makes sense. The one thing I'm still unclear about is how to create `Oid`s for files that are modified in the working tree, but not staged. This is why I thought I needed to call `update_all` -- so that the index would stage those changes and make `Oid`s available.

My assumption is that if a change is staged, then it should have an index entry with an Oid. But, I'm not sure how to get from unstaged -> staged without interacting with the index.

(I'm not super familiar with Git internals, so apologies if this question is misguided.)

Reply:
Ah, on the specific question of file -> OID, it looks like `Repo::blob_path` would work

Reply:
We don't want to stage the changes as part of the commit process, or use an `Index` structure to ultimately make the commit. To get the OIDs for the changed files, you can use `Repo::blob_path` like you mentioned. To make the commit object, you'll want to take the current commit's tree and modify it by inserting the changed entries. You can use this crate's utility function `tree::hydrate_tree` to do that. Then pass the tree to this crate's `Repo::create_commit`.